### PR TITLE
Add message types for HotShot upgradability

### LINKED ISSUE #2372: [Upgradability] Create a new certificate type
### What is this task and why do we need to work on it?

There should be a new certificate type that we use to collect signatures for a decision to upgrade the network.

### What work will need to be done to complete this task?

Implement the certificate type.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

A new certificate type is implemented. 

This issue does not yet require the certificate design to be finalized, or fully integrated in HotShot.

### Branch work will be merged to (if not the default branch)

_No response_

### LINKED ISSUE #2383: [Upgradability] Add a version field to the start of messages
### What is this task and why do we need to work on it?

To allow network protocol upgrades, messages need an identifier for the protocol version used in the message. We would like to add a protocol version number at the start of every serialized message.

### What work will need to be done to complete this task?

Add version numbers to the start of every serialized message.

### What are the acceptance criteria to close this issue?

A version number field is implemented with tests, and existing tests pass.

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by ss-es**
Closes #2372, closes #2383

### This PR: 
Adds message/certificate types for upgrade proposals.

### This PR does not: 
Add any logic for handling upgrade proposals, or handle upgrades in general.

### Key places to review: 
- Names for message/certificate types: e.g. is 'Proposal' more appropriate?
- Serialization test: is there a better place to put it than into a new crate?

### CODE REVIEW

File: crates/task-impls/src/events.rs [hunk 1]
```diff
@@ -111,4 +111,6 @@ pub enum HotShotEvent<TYPES: NodeType> {
     ///
     /// Like [`HotShotEvent::DAProposalRecv`].
     VidDisperseRecv(Proposal<TYPES, VidDisperse<TYPES>>, TYPES::SignatureKey),
+    /// Upgrade certificate has been received from the network.
+    UpgradeCertificateRecv(UpgradeCertificate<TYPES>),
```
Comment (shenkeyao) [inline crates/task-impls/src/events.rs:115]: Do we also need `UpgradeCertificateSend`, `UpgradeVoteSend`, and `UpgradeVoteRecv`?

File: crates/task-impls/src/network.rs [hunk 2]
```diff
@@ -85,6 +86,9 @@ impl<TYPES: NodeType> NetworkMessageTaskState<TYPES> {
                             GeneralConsensusMessage::TimeoutVote(message) => {
                                 HotShotEvent::TimeoutVoteRecv(message)
                             }
+                            GeneralConsensusMessage::UpgradeCertificate(message) => {
+                                HotShotEvent::UpgradeCertificateRecv(message)
+                            }
```
Comment (shenkeyao) [inline crates/task-impls/src/network.rs:91]: Related to the comment in events.rs, should we also handle the other vote/cert events here?

File: crates/unit-tests/Cargo.toml
Comment (shenkeyao) [inline crates/unit-tests/Cargo.toml:1]: Would it be better to move the unit tests to the `testing` crate, so we won't have two testing crates that may cause confusion later?

**Review by shenkeyao — COMMENTED** [review]
Looks good! Left some comments about additional event types and where to put unit tests.

File: crates/task-impls/src/events.rs (hunk 1 above)
Comment (ss-es) [reply crates/task-impls/src/events.rs:123]: Yes, I added these!

File: crates/task-impls/src/network.rs (hunk 2 above)
Comment (ss-es) [reply crates/task-impls/src/network.rs:91]: Yes - added!

File: crates/unit-tests/Cargo.toml
Comment (ss-es) [reply crates/unit-tests/Cargo.toml:1]: Yes - moved!

File: crates/types/src/version.rs
Comment (shenkeyao) [inline crates/types/src/version.rs:1]: I was just checking my understanding of structs in this file. Are we going to add more version structs to this file representing future versions? And are the current two structs placeholders to which we'll add fields later?

**Review by shenkeyao — COMMENTED** [review]
Just a few comments. Looks good overall!

File: crates/types/src/message.rs [hunk 3]
```diff
@@ -32,6 +33,11 @@ use std::{fmt::Debug, marker::PhantomData};
 #[derive(Serialize, Deserialize, Clone, Debug, Derivative, PartialEq, Eq, Hash)]
 #[serde(bound(deserialize = "", serialize = ""))]
 pub struct Message<TYPES: NodeType> {
+    /// The major version number of the protocol in use for this message
+    pub major_version: u32,
+    /// The minor version number of the protocol in use for this message
+    pub minor_version: u32,
```
[the remark this answers is not in the document]
Comment (ss-es) [reply crates/types/src/message.rs]: Yes - good catch! I realized I was defining versions in multiple places which caused this mistake. I made the Version a common struct to fix this!

File: crates/types/src/version.rs
Comment (ss-es) [reply crates/types/src/version.rs:1]: Oops, that file was meant for another branch -- it was not meant to be in this PR, so I removed it!

File: crates/types/src/traits/consensus_type.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (elliedavidson) [inline crates/types/src/traits/consensus_type.rs:1]: Can we remove this file?

**Review by elliedavidson — CHANGES_REQUESTED** [review]

**Review by elliedavidson — APPROVED** [review]

**Pull request merged**